### PR TITLE
Add documentation for ip pool service allocation

### PR DESCRIPTION
**Pull request by pperiyasamy**
This adds web content for ip pool with service allocation feature and a sample yaml file for such configuration.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

### CODE REVIEW

File: website/content/configuration/_advanced_ipaddresspool_configuration.md [hunk 1]
```diff
@@ -48,6 +48,52 @@ To specify a single IP address in a pool, use `/32` in the CIDR notation
...
+          foo: bar
+    serviceSelectors:
+      - matchExpressions:
+          - {key: app, operator: In, values: [bar]}
+```
+
+The above IPAddressPool example is pinned to service(s) with has a label
+matching expression `key: app, operator: In, values: [bar]` and created
+either in `namespace-a` or `namespace-b` or any other namespace(s) which
+has a matching label `foo: bar`.
```
Comment (fedepaol) [inline website/content/configuration/_advanced_ipaddresspool_configuration.md:95]: I'd also add a note to say that if a service chooses a pool (or an ip belonging to a pool) which doesn't match the service will stay in pending.

File: website/content/configuration/_advanced_ipaddresspool_configuration.md [hunk 2]
```diff
@@ -48,6 +48,52 @@ To specify a single IP address in a pool, use `/32` in the CIDR notation
...
+      - matchLabels:
+          foo: bar
+    serviceSelectors:
+      - matchExpressions:
+          - {key: app, operator: In, values: [bar]}
+```
+
+The above IPAddressPool example is pinned to service(s) with has a label
+matching expression `key: app, operator: In, values: [bar]` and created
+either in `namespace-a` or `namespace-b` or any other namespace(s) which
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md]: yes, fixed with this commit https://github.com/metallb/metallb/pull/1821/commits/cd7f7fb5588e54cac51322580c787f61af1494ac.

File: website/content/configuration/_advanced_ipaddresspool_configuration.md [hunk 3]
```diff
@@ -48,6 +48,52 @@ To specify a single IP address in a pool, use `/32` in the CIDR notation
...
 {{% /notice %}}
 
+### Reduce scope of address allocation to specific Namespace and Service
+
+This option can be used to reduce the scope of particular IPAddressPool
+to set of namespaces and services, by adding an optional namespace/service
+selector.
+
+The use the pool with lowest value of priority in case of multiple matches.
+This is deterministic and will make the behaviour more clear to the users.
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md]: removed.

File: website/content/configuration/_advanced_ipaddresspool_configuration.md (hunk 1 above)
Comment (pperiyasamy) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md]: done.

File: internal/allocator/allocator.go [hunk 4]
```diff
@@ -345,13 +337,57 @@ func (a *Allocator) sortedPoolsForService(svc *v1.Service) []*config.Pool {
...
 func (a *Allocator) isPoolCompatibleWithService(p *config.Pool, svc *v1.Service) bool {
-	if p.ServiceAllocations != nil && p.ServiceAllocations.Namespaces.Len() > 0 &&
-		!p.ServiceAllocations.Namespaces.Has(svc.Namespace) {
-		return false
+	svcLabels := labels.Set(svc.Labels)
+	if p.ServiceAllocations != nil && p.ServiceAllocations.Namespaces.Len() > 0 {
+		if !p.ServiceAllocations.Namespaces.Has(svc.Namespace) {
+			return false
+		}
+		hasSvcLabelMatch := false
```
Comment (fedepaol) [inline internal/allocator/allocator.go:378]: Is this necessary? If we don't return false here we'll fall back in the check below, and if the service doesn't match we'll return false anyway

File: internal/allocator/allocator.go [hunk 5]
```diff
@@ -310,17 +310,9 @@ func (a *Allocator) sortedPoolsForService(svc *v1.Service) []*config.Pool {
...
-				continue
-			}
-			if nsPool.ServiceAllocations.Priority > 0 {
-				priorityPools = append(priorityPools, nsPool)
-				continue
-			}
-			noPriorityPools = append(noPriorityPools, nsPool)
-		}
+	if len(a.pools.ByNamespace[svc.Namespace]) > 0 {
+		priorityPools, noPriorityPools = a.sortedNamespacePoolsForService(svc)
```
Comment (fedepaol) [inline internal/allocator/allocator.go:314]: thinking again about this, now the priority / no priority is useless as long as we handle the compare function properly, because at the end we exit from here with one single list

Comment (fedepaol) [reply internal/allocator/allocator.go:314]: Also, instead of reimplementing it, I'd leverage the isPoolCompatibleWithService in both this loop and the one below. Basically the idea is "we see that this matches, let's see if it's compatible.

I'd do the same in the loop below even if it's logically reduntant now and burns a few cycles, because in the future we might extend the "isPoolCompatible" function with more checks.

File: internal/allocator/allocator_test.go [hunk 6]
```diff
@@ -520,6 +531,29 @@ func TestAssignment(t *testing.T) {
 			ips:     []string{"1.2.6.10"},
 			wantErr: true,
 		},
+		{
+			desc:   "attempt to assign ip from pool pinned with namespace and service label",
```
Comment (fedepaol) [inline internal/allocator/allocator_test.go:535]: can we add two tests that should fail because the labels are compatible but the ns is not and viceversa?

File: internal/config/config.go [hunk 7]
```diff
@@ -561,7 +561,8 @@ func poolsByNamespace(pools map[string]*Pool) map[string][]string {
 func poolsByServiceSelector(pools map[string]*Pool) []string {
 	var poolsByServiceSelector []string
 	for _, pool := range pools {
-		if pool.ServiceAllocations == nil || len(pool.ServiceAllocations.ServiceSelectors) == 0 {
+		if pool.ServiceAllocations == nil || len(pool.ServiceAllocations.Namespaces) > 0 ||
```
Comment (fedepaol) [inline internal/config/config.go:564]: is this change necessary? I think we should avoid the optimization of "checking both first, serviceselector only later" because it might bite us in the future. I guess it's cleaner if we have two separate sets, and each time we check for the whole compatiibility.

File: website/content/configuration/_advanced_ipaddresspool_configuration.md [hunk 8]
```diff
@@ -48,6 +48,42 @@ To specify a single IP address in a pool, use `/32` in the CIDR notation
...
+          foo: bar
+    serviceSelectors:
+      - matchExpressions:
+          - {key: app, operator: In, values: [bar]}
+```
+
+The above IPAddressPool example is pinned to service(s) with has a label
+matching expression `key: app, operator: In, values: [bar]` and created
+either in `namespace-a` or `namespace-b`.
+When a service explicitly chooses an IPAddressPool (or an IP Address
```
Comment (fedepaol) [inline website/content/configuration/_advanced_ipaddresspool_configuration.md:83]: Nit: add the note modifier here (just because the pr needs to be reworked anyway)

Comment (pperiyasamy) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md:85]: done.

File: internal/allocator/allocator_test.go (hunk 6 above)
Comment (pperiyasamy) [reply internal/allocator/allocator_test.go:546]: done

File: internal/allocator/allocator.go (hunk 4 above)
Comment (pperiyasamy) [reply internal/allocator/allocator.go:377]: yes, this check is needed for service allocations which has only namespaces, but not service selectors.

File: internal/allocator/allocator.go (hunk 5 above)
Comment (pperiyasamy) [reply internal/allocator/allocator.go]: yes, correct, we don't need `noPriorityPools` and it can be done with sort function itself.
for other comment, though these methods has some duplicate piece of code and each one is used for different purpose. I think existing structure looks to be cleaner.

File: internal/config/config.go (hunk 7 above)
Comment (pperiyasamy) [reply internal/config/config.go]: ok, reverted it.

File: internal/allocator/allocator.go (hunk 4 above)
Comment (fedepaol) [reply internal/allocator/allocator.go:377]: but if a service has no service selectors we'll go to the final return, right?

File: internal/allocator/allocator.go (hunk 5 above)
Comment (fedepaol) [reply internal/allocator/allocator.go]: I think having two loops (one for namespaces and the other for services) that check the match on each side, and then the compatibility is more symmetric and easier to reason about. 

Every loop should be like "check if this pool matches upon criteria x, and check if matches completely". This is easier to follow because we don't need to think that we already checked the namespaces in the previous loop, and  will make things easier if we add a third criteria.


So, what I am proposing is to use the `isPoolCompatibleWithService` function while we check in `getNamespacePoolsForService` and also in the `range a.pools.ByServiceSelector` loop. Also, since the former will be slimmer because of that, we can consider either removing the wrapper function (or wrapping the service selector logic in another function too, it feels asymmetric).

Comment (pperiyasamy) [reply internal/allocator/allocator.go]: OK, I see your point. changed it accordingly, Hope it looks better now.

File: website/content/configuration/_advanced_ipaddresspool_configuration.md (hunk 8 above)
Comment (fedepaol) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md:85]: what I meant is, use the `{{% notice note %}}` wrapper to highlight this.

Comment (fedepaol) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md:85]: and since we are going to touch this, I'd rephrase with When a service explicitly chooses an IPAddressPool via the xxx annotation or an ip via the spec.loadbalancerip...

File: internal/allocator/allocator.go (hunk 5 above)
Comment (fedepaol) [reply internal/allocator/allocator.go]: yes! Much better

**Comment by fedepaol** [comment]
one final nit about the doc part, and lgtm

File: internal/allocator/allocator.go [hunk 9]
```diff
@@ -305,44 +305,29 @@ func (a *Allocator) Allocate(svcKey string, svc *v1.Service, serviceIPFamily ipf
...
 // priority ip pools.
-func (a *Allocator) sortedPoolsForService(svc *v1.Service) []*config.Pool {
-	var priorityPools, noPriorityPools []*config.Pool
+func (a *Allocator) pinnedPoolsForService(svc *v1.Service) []*config.Pool {
+	var pools []*config.Pool
 	if svc == nil {
-		return priorityPools
+		return pools
 	}
 	for _, nsPoolName := range a.pools.ByNamespace[svc.Namespace] {
```
Comment (oribon) [inline internal/allocator/allocator.go:313]: why isn't it enough to just iterate over all of the pools ByName instead of multiple fors? the `isPoolCompatible` would filter out the unrelated pools anyways? (sorry if I'm confusing stuff, not familiar with this part and the Pools struct)

Comment (fedepaol) [reply internal/allocator/allocator.go:313]: yep, but here we save some loops by skipping the non pinned loops, as "byName" contains all of them

Comment (oribon) [reply internal/allocator/allocator.go:313]: not sure about saving loops, as it seems the current implementation can add the same pool twice to the `pools` slice. I think it would be more readable to just iterate once over everything and the performance hit (if at all) shouldn't be too big

Comment (pperiyasamy) [reply internal/allocator/allocator.go:313]: we can avoid pool duplicates in the `pools` slice with a change in config.go, but that was reverted based on this [comment](https://github.com/metallb/metallb/pull/1821#discussion_r1107011002). I think iterating over all pools may be somewhat costlier than current one as it needs to validate ns and svc labels for every ns/svc pools.

Comment (pperiyasamy) [reply internal/allocator/allocator.go:313]: oh! no, the isPoolCompatibleWithService ensures no duplicates get added into `pools` slice, i think we're better here.

Comment (oribon) [reply internal/allocator/allocator.go:313]: what I meant (and I might be wrong), that in the current way byNS/bySelector might overlap (or contain everything?) and that way we aren't really saving any iterations by not going once over everything

File: internal/allocator/allocator.go [hunk 10]
```diff
@@ -374,6 +359,21 @@ func (a *Allocator) Pool(svc string) string {
 	return ""
 }
 
+func sortPools(pools []*config.Pool) {
+	sort.Slice(pools, func(i, j int) bool {
+		if pools[i].ServiceAllocations.Priority > 0 &&
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply internal/allocator/allocator.go:367]: added comment now.

File: website/content/configuration/_advanced_ipaddresspool_configuration.md [hunk 11]
```diff
@@ -48,6 +48,44 @@ To specify a single IP address in a pool, use `/32` in the CIDR notation
 (e.g. `42.176.25.64/32`).
 {{% /notice %}}
 
+### Reduce scope of address allocation to specific Namespace and Service
+
+This option can be used to reduce the scope of particular IPAddressPool
+to set of namespaces and services, by adding an optional namespace
+and/or service selectors.
+
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md:59]: ok, done

File: website/content/configuration/_advanced_ipaddresspool_configuration.md [hunk 12]
```diff
@@ -48,6 +48,44 @@ To specify a single IP address in a pool, use `/32` in the CIDR notation
...
+kind: IPAddressPool
+metadata:
+  name: ippool-ns-service-alloc-sample
+  namespace: metallb-system
+spec:
+  addresses:
+    - 192.168.20.0/24
+  avoidBuggyIPs: true
+  serviceAllocation:
+    priority: 50
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md:71]: done [here](https://github.com/metallb/metallb/pull/1821/files#diff-02a9536c1258b76a7825571c2016dd4807f0f8ffb66485d5fb3260eb5b01d806R88-R94).

File: internal/allocator/allocator.go (hunk 9 above)
Comment (pperiyasamy) [reply internal/allocator/allocator.go:313]: byNS/bySelector may contain subset or all of ip pools. but it depends on user's pool config. If we do it by iterating over all pools via `a.pools.ByName`, the lookup cost for desired ns/svc pools are high. @fedepaol your thoughts ?

File: website/content/configuration/_advanced_ipaddresspool_configuration.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (oribon) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md:70]: I think it should be documented here as well for the users + the fact that not specifying a priority / setting priority 0 is the lowest priority.

File: internal/allocator/allocator.go (hunk 9 above)
Comment (oribon) [reply internal/allocator/allocator.go:313]: iiuc, if most of the people won't use this feature the byNS/bySelector will always contain all of the pools.
having said that, I wouldn't want this to block this PR if we intend to have it in the next release as it's not too crucial.

Comment (fedepaol) [reply internal/allocator/allocator.go:313]: I can't think about a scenario where there are more than... 100 pools, to say a number.
On the other hand, I think the current version is more than understandable so I'd say let's go with this version, I don't mind cleaning up later if it becomes too messy / if we add another discriminator.

File: website/content/configuration/_advanced_ipaddresspool_configuration.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (fedepaol) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md:70]: :+1: , and I think this is the last standing comment right?

[the remark this answers is not in the document]
Comment (oribon) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md:70]: yep, other than that the pr looks good

[the remark this answers is not in the document]
Comment (pperiyasamy) [reply website/content/configuration/_advanced_ipaddresspool_configuration.md:70]: added [here](https://github.com/metallb/metallb/pull/1821/files#diff-02a9536c1258b76a7825571c2016dd4807f0f8ffb66485d5fb3260eb5b01d806R92-R94).
